### ISSUE TITLE
Improve error handling

### ISSUE THREAD

**Issue by lucamlouzada** [gentzkow/GentzkowLabTemplate#18]
This issue is part of an effort to implement substantive improvements to the lab template, as discussed in #16.

In this issue, the goal is to work on error handling. The main points to be addressed per the [plans for next steps](https://github.com/gentzkow/GentzkowLabTemplate/blob/c12f4a1dbd2928d762a4b8377561c09dfc488028/issue/nextstep.lyx) are:

- Make.sh should report errors in Python / R / Stata / Latex scripts.
- Improve logging -- make.log should nicely report the start and end of make logging, the start and end of each script, and as much verbose feedback on what's going on within each script as possible.
- Error tracking should happen in the run command. It should be reported both in the terminal and in the log file.
- Detect and report all kinds errors, not only errors when running a program, but also the errors that stops the program from running and even when program is not available via the command line.
- If the desired output have not been generated but the program still runs well, report an error.

I am assigning myself to work on this.

**Comment by lucamlouzada** [comment]
I pushed some significant changes to each of the run_xx scripts, as well as to the make.sh scripts. The main changes and the reasoning behind them are as follows:
### Run_xx scripts
- The previous logic of the shell scripts in the lines beginning with "# run program, add output to logfile" and "# report on errors" were not appropriately directing error messages to the log file. I changed this logic so now we store the output and errors in a variable and then treat this variable accordingly, showing only clean messages and warnings in the terminal and leaving all the detailed information to the log files. Note that for run_stata, this approach does not work and thus it is necessary to [parse the stata .log](https://gist.github.com/pschumm/b967dfc7f723507ac4be) files to identify errors appropriately. 
- I added a specific error check to the run_xx scripts to let users know if the programs are missing from the command line
- I added an output check to the run_xx scripts. In the beginning and end of the code, the scripts capture the state of the output folder and the script then lists the files that were created
- I added a timestamp message that records the beginning and ending of each script
- I changed the way some comments were written to make the scripts more standardized and made minor changes such as adding "\n" to the beginning of some lines to that the messages displayed in the terminal and the log are cleaner

### Make.sh
- The previous code included "set -e" in the beginning which means the script would exit upon any errors. However, this prevents us from making more detailed error handling (if this is activated, the error handling specified in the run_xx scripts does not work because when there are errors make.sh just exists the code). Therefore, I added a specific error handler function to capture errors that happen in the shell script itself, while errors related to the run_xx scripts are dealt within them as explained above. The error handler redirects the errors to the log file, so it is only defined after the log file has been created - "set -e" is active for the beginning of the code, in case there are errors in the first line of the shell script, then it is deactivated and the error handler is activated.
- The previous make.sh script ran all scripts in a subshell. I am not entirely sure of the reason why (the comment said this was so it would return to the other working directory after running). However, this makes it more difficult to deal with errors within the subshell, so I removed this aspect and made the script simpler by just setting the working directory back manually after running the code

**Comment by lucamlouzada** [comment]
I just added two pushes to this branch to reflect a discussion with @gentzkow in which we noticed some errors when running the scripts from a `zsh` shell instead of `bash`.  There seems to be some incompatibilities due to small differences in syntax such that some commands yield different outputs in `zsh` and `bash`. With the small changes I made, the outputs should be the same. I will add a step in the suggested review framework in the PR #19 to suggest testing in `zsh` as well.

**Comment by lucamlouzada** [comment]
I have investigated the discrepancies between `zsh` and `bash`, and they are related to the fact that exceptions are inherited from functions differently in each of them. In 888557d, I added a line to unset the error trap within the `run_xx` scripts. This allows the error handling within these scripts to be identified by the parent shell script. 

It is still the case that errors in `my_shell_script.sh` will trigger the outer shell error handling in`zsh` and not in `bash` - see screenshot below - the second warning, which captures shell errors, shows up only in the `zsh` script but not in the `bash` script)
<img width="617" alt="image" src="https://github.com/user-attachments/assets/95076d38-9b7b-4a9f-b0ba-df6afd3b143a">

Hence, the only difference that we should observe is that when calling shell scripts that have errors within `make.sh` using the `run_shell.sh` function, there will be an extra warning  in `zsh` compared to `bash` as in the screenshot above. Note that this behavior is not observed for error in other kinds of scripts - in python for example the behavior is consistent:
<img width="570" alt="image" src="https://github.com/user-attachments/assets/9cb5b212-3a7c-4ba4-b73a-d9aacbfb6e22">

I have added a step in the review framework in #19 to test both in `zsh` and `bash.`

**Comment by lucamlouzada** [comment]
**Summary:**
In this issue, we implemented several changes to the `run_xxx` and `make.sh` scripts to improve error handling. The main changes were related to error logging in the `run_xxx` scripts and the addition of an error trap in `make.sh`. More details about the changes can be found in https://github.com/gentzkow/GentzkowLabTemplate/issues/18#issuecomment-2359557206, https://github.com/gentzkow/GentzkowLabTemplate/issues/18#issuecomment-2380285621, and in the associated PR #19. 

Changes were merged to master in eda0569. The final state of the issue branch is [here](https://github.com/gentzkow/GentzkowLabTemplate/tree/4e3759016bb299449f799466488c0077b31f1a1e). 

Thanks @ShiqiYang2022 for the thorough review and feedback, as well as @gentzkow for great suggestions.
